sockeye/rnn.py

```diff
@@ -295,6 +302,47 @@ def __call__(self, inputs, states):
         return next_h, [next_h, next_c]
 
 
+class RecurrentDropoutLSTMCell(mx.rnn.LSTMCell):
+    """
+    LSTMCell with recurrent dropout without memory loss as in:
+    http://aclanthology.coli.uni-saarland.de/pdf/C/C16/C16-1165.pdf
+    """
+
+    def __init__(self, num_hidden, prefix='lstm_', params=None, forget_bias=1.0, dropout: float = 0.0) -> None:
+        super().__init__(num_hidden, prefix, params, forget_bias)
+        self.dropout = dropout
+
+    def __call__(self, inputs, states):
+        self._counter += 1
+        name = '%st%d_' % (self._prefix, self._counter)
+        i2h = mx.sym.FullyConnected(data=inputs, weight=self._iW, bias=self._iB,
+                                    num_hidden=self._num_hidden * 4,
+                                    name='%si2h' % name)
+        h2h = mx.sym.FullyConnected(data=states[0], weight=self._hW, bias=self._hB,
+                                     num_hidden=self._num_hidden * 4,
+                                     name='%sh2h' % name)
+        gates = i2h + h2h
+        slice_gates = mx.sym.SliceChannel(gates, num_outputs=4,
+                                          name="%sslice" % name)
+        in_gate = mx.sym.Activation(slice_gates[0], act_type="sigmoid",
+                                    name='%si' % name)
+        forget_gate = mx.sym.Activation(slice_gates[1], act_type="sigmoid",
+                                        name='%sf' % name)
+        in_transform = mx.sym.Activation(slice_gates[2], act_type="tanh",
+                                         name='%sc' % name)
+        if self.dropout > 0.0:
```

Comment:
As this is a cell explicitly including dropout (and the test for >0.0 is also done outside), it may make sense to make this a precondition in the constructor.

Reply:
good point, done